kep/690-gpu-topology-aware-scheduling/README.md

```diff
@@ -0,0 +1,473 @@
+
+This is how the architecture generally looks like, as you see, we need several components to work together.
+
+- Kube-scheduler: with new implemented plugin to help finding the most suitable node with satisfied GPUs
+- kubelet & k8s-device-plugin: work as before
+- [GFD](https://github.com/NVIDIA/gpu-feature-discovery) & [NFD](https://github.com/kubernetes-sigs/node-feature-discovery): patch GPU topology matrix to Node annotation, currently, it's not supported yet, see <https://github.com/NVIDIA/k8s-device-plugin/issues/465>,
+we'll help to implement this.
+
+**How to pick the best effort GPUs in one node**
+
+This is implemented by default inside k8s-device-plugin, which leverages the library of [go-gpuallocator](https://github.com/NVIDIA/go-gpuallocator).
+Generally, go-gpuallocator will try to find a set of GPUs with the highest bandwidth. The bandwidth is scored by the link type,
+see details [here](https://github.com/NVIDIA/go-gpuallocator/blob/8fc3087147de2bf654ee91929e9a70d754c6bae1/gpuallocator/besteffort_policy.go#L316-L367).
+
+Note that go-gpuallocator will not do actual allocation, this is what k8s-device-plusin does. go-gpuallocator will only dry-run
+to find the best set of GPUs.
+
+The allocation algorithm is:
+
+- We'll have a map of P2P link type to Score, like `SingleNVLINKLink` values 100 scores,`TwoNVLINKLinks` values 200 scores.
+- When we require X GPUs for Pod, we'll get M possibles partitions with size of X. And calculate the total scores of each partitions.
+For example, we have GPU0, GPU1, GPU2, GPU3 and request 3 GPUs, we'll have Partition0[[GPU0, GPU1, GPU2], [GPU3, nil, nil]],
+Partition1[[GPU0, GPU1, GPU3], [GPU2, nil, nil]], Partition2[[GPU1, GPU2, GPU3], [GPU0, nil, nil]], the score of Partition0 is
+sum(score(GPU0, GPU1), score(GPU0, GPU2), score(GPU1, GPU2), score(GPU3, nil)).
+- Assume Partition0 has the highest score, then we'll pick a set of the partition who has the highest score. Here is the [GPU0, GPU1, GPU2].
+
+**How to pick the most suitable Node across cluster**
+
+Kube-scheduler will follow the same algorithm of go-gpuallocator, it will process all possible nodes and
+calculate the scores of each Node.
+
+For scheduler-plugin, we'll implement `PreScore` and `Score` extension points:
+
+- PreScore: If Pod doesn't require `nvidia.com/gpu`, return SKip directly.
+- Score: reuse the go-gpuallocator's scoring algorithm to get the original score, then for normalizing,
+
+    ```golang
+    maxScore = limitQuota * (limitQuota - 1) * 1800 / 2 // 1800 is the maximum score between two GPUs, on behalf of 18 NVLinks
+    finalScore = originalScore * 100 / maxScore
+    ```
+
+A potential issue here is if the score of the link type is too small, like `P2PLinkCrossCPU` maps to 10, then we may loss the precision here
+and got 0 score finally.
+
+**In summary**
+
+Based on the same scoring algorithm, kube-scheduler will find the best-fit node, but no need to tell the device plugin which GPUs are they,
```

Comment:
IIIUC, kube-scheduler reuses NVIDIA/go-gpuallocator algorithm to calculate the best GPU device id for allocation and score the nodes accordingly. But scheduler can't 100% sure which GPU will be used eventually, it "guesses" the result depends on the topology information reported by NFD, and device plugin will make its own decision, am I right? 
If so, does scheduler get to know which GPU has already been used?

Reply:
Missed that part, will add it. Thanks.